idaes/apps/uncertainty_propagation/uncertainties.py

```diff
@@ -0,0 +1,248 @@
+    # get gradient of the objective function, constraints, and the column number of each theta
+    gradient_f, gradient_c, col, line_dic = get_dfds_dcds(model, theta_names, tee)
+    dsdp, col  = get_dsdp(model, theta_names, theta, var_dic,tee)      
+    
+    gradient_f_dic = {}
+    for i in range(len(col)):
+        gradient_f_dic["d(f)/d("+col[i]+")"] = gradient_f[i]
+    num_constraints = len(list(model.component_data_objects(Constraint,
+                                                            active=True,
+                                                            descend_into=True)))
+    gradient_c_dic = {}
+    if num_constraints > 0 :
+        gradient_c = np.array([i for i in gradient_c if not np.isclose(i[2],0)])
+        row_number, col_number = np.shape(gradient_c)
+        gradient_c_dic = {}
+        for i in range(row_number):
+            gradient_c_dic["d(c"+ str(int(gradient_c[i,1]))+")/d("+col[int(gradient_c[i,0]-1)]+")"] = gradient_c[i,2]
+    dsdp_dic = {}
+    for i in range(len(theta_names)):
+        for j in range(len(col)):
+            if SensitivityInterface.get_default_block_name() not in col[j]:
+                dsdp_dic["d("+col[j] +")/d("+theta_names[i]+")"] =  -dsdp[i, j]
+    x_list = [x for x in col if x not in var_dic.keys()] #x_list includes only decision variables (exclude theta_names) 
+    p_list = theta_names #This makes cov and fxxp have the same order 
+    fxxp = []
+    for i in p_list:
+        fxxp_tmp = 0
+        for j in x_list:
+            fxxp_tmp = fxxp_tmp + gradient_f_dic['d(f)/d('+j+')']*dsdp_dic['d('+j+')/d('+i+')']
+        fxxp.append(fxxp_tmp)
+    fxxp = np.array(fxxp)
+    # calculate error propagation df/dp*cov_p*df/dp + (df/dx*dx/dp)*cov_p*(df/dx*dx/dp)
+    # objective function: 
+    for cc in list(model.component_data_objects(Objective,
+                                            active=True,
+                                            descend_into=True)):
+        # save gradient of the objective with respect to only theta_names
+        if any(ele in str(cc.expr) for ele in theta_names):
+            gradient_f_theta = gradient_f[[int(line_dic[v]) -1 for v in theta_names]]
+    propagation_f = {}
+    # calculate error propagation of the objective fuction
+    if 'gradient_f_theta' in locals():
+        propagation_f = np.dot(gradient_f_theta,np.dot(cov,np.transpose(gradient_f_theta))) + np.dot(fxxp,np.dot(cov,np.transpose(fxxp)))
+    else:
+        propagation_f = np.dot(fxxp,np.dot(cov,np.transpose(fxxp)))
+    # constraints: 
+    # calculate error propagation dc/dp*cov_p*dc/dp + (dc/dx*dx/dp)*cov_p*(dc/dx*dx/dp)
+    num_constraints = len(list(model.component_data_objects(Constraint,
+                                                            active=True,
+                                                            descend_into=True))) 
+    if num_constraints > 0:
+        # save constraints into gradient_cc only if constraints includes theta 
```

Comment:
Why reduce this to only the constraints that include theta? Why not keep the larger matrix with lost of zero entries? This is easy to do with a sparse numpy array. The benefit is that we do not need to keep track of the constraints that contain only theta.

Reply:
I updated it to include all values in a sparse array.